repose-aggregator/components/services/datastore-service/datastore-service-impl/src/test/scala/org/openrepose/nodeservice/distributed/servlet/DistributedDatastoreServletTest.scala

```diff
@@ -0,0 +1,93 @@
+/*
+ * _=_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=
+ * Repose
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Copyright (C) 2010 - 2015 Rackspace US, Inc.
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+package org.openrepose.nodeservice.distributed.servlet
+
+import java.net.InetAddress
+import java.util.Collections
+import javax.servlet.http.HttpServletResponse
+import javax.servlet.http.HttpServletResponse._
+
+import io.opentracing.Tracer
+import io.opentracing.mock.MockTracer
+import org.junit.runner.RunWith
+import org.mockito.Matchers.anyString
+import org.mockito.Mockito
+import org.mockito.Mockito.verify
+import org.openrepose.core.services.datastore.distributed.ClusterConfiguration
+import org.openrepose.core.services.datastore.distributed.config._
+import org.openrepose.core.services.datastore.impl.distributed.CacheRequest.CACHE_URI_PATH
+import org.openrepose.core.services.datastore.{Datastore, DatastoreAccessControl, DatastoreService}
+import org.scalatest.junit.JUnitRunner
+import org.scalatest.mock.MockitoSugar
+import org.scalatest.{BeforeAndAfterEach, FunSpec, Matchers}
+import org.springframework.mock.web.{MockHttpServletRequest, MockHttpServletResponse}
+
+
+@RunWith(classOf[JUnitRunner])
+class DistributedDatastoreServletTest extends FunSpec with BeforeAndAfterEach with Matchers with MockitoSugar {
+
+  private val mockDatastoreService = mock[DatastoreService]
+  private val mockDatastore = mock[Datastore]
+  Mockito.when(mockDatastoreService.getDefaultDatastore).thenReturn(mockDatastore)
+  private val distributedDatastoreConfiguration = mock[DistributedDatastoreConfiguration]
+  val tracerSpy = Mockito.spy(new MockTracer)
```

Comment:
Personally, I am a fan of re-creating mocks for each test to ensure that there is no test bleed. The pattern I have started following is to declare all of my variables as `var`s then assign mocks to those variables in the `beforeEach` method. For example:
```
var datastoreService: DatastoreService = _
var datastore: Datastore = _

override def beforeEach(): Unit = {
  datastoreService = mock[DatastoreService]
  datastore = mock[Datastore]

  when(mockDatastoreService.getDefaultDatastore).thenReturn(mockDatastore)
}
```

You could, of course, use `val`s and `Mockito.reset(...)` instead of `var`s and `new`s, but I would rather rely on the language doing the right thing than the `Mockito` library. Although, there is a slight overhead to doing it my way.

I would also recreate the class under test for each test for the same reason (unless I am testing a singleton or static methods on a class).

Interestingly, you have already done this with the request and response.

Reply:
This is one of those we either trust what Mockito/OpenTracing provides or we don't. If we do, then I don't see an issue using their provided API. If we don't, then let's find something else.

That said, I typically use the pattern of if it is intended to be a long lived entity (e.g. a `Servlet`), then I like to test it under those conditions so we have a better chance of possibly finding the odd ball scenarios that cause a failure. As always it is never a guarantee, but it is more accurate testing for it's intended lifecycle.

That is also the intent behind the creation of the request and response every time. These are short lived one off entities and this is in line with their lifecycle.

In the end, I did remove the `Spy`, but left the `reset` on the `MockTracer` and the`DistributedDatastoreServlet` remains long lived as this is their intended lifecycle.

Reply:
Your intended lifecycle point is an interesting one. I almost want to agree, but as I think about it more, I don't think I do. The problem I have is that while your tests may match the lifecycle of the object under tests, they also allow state/behavior to leak between tests. That makes individual tests less reliable since leftover state from a different test has the potential to change the behavior in the current test. That means that something like changing the ordering of tests might change the results, which does not seem good. Additionally, it becomes more difficult to reason about tests since once again you may have to account for leakage. If every test is self-contained, reasoning about the test is much more straightforward.

Now, a counter argument would be to say that tests failing because of test bleed is a good thing because it tells us when there is bleed. However, as noted above, that bleed may not be trivial to track down even with the tests. It also may not be an issue in the production code, but instead, an issue in the test code. And, of course, there is the fact that these are unit tests. Unit tests are only meant to test a unit of work, and in my mind, the best way to do so is to isolate the unit in question so that I can identify/specify the exact causality for some behavior.

As for the `Mockito.reset(...)` vs `var`/`new` thing, sure, either way is fine. It's less that I don't trust `Mockito.reset(...)` and more than there's no need to trust `Mockito.reset(...)`. To me, `Mockito.reset(...)` seems like a performance optimization to `var`/`new`, but as it turns out, the incredibly minor performance gain we get from it does not really bother me when it comes to tests. `var`/`new` is a pattern than works for non-mocked objects as well, so there's a consistency argument there. Ultimately, though, unless you're using `reset(...)` within a test (which is a test smell!), I don't overly care.

Just my 2c. As it stands, these tests should meet what we want to get out of them, so if you still want to leave them as-is, that's fine.

Reply:
I too agree, the extremely minimal cost of the `val`/`reset()` vs `var`/`new` shouldn't be a concern in these tests.

I've been thinking some more on this and your point about the state/behavior leakage is a fair one. As they were written, these tests were more in line with this projects use of Functional Integration Tests. That said, I've updated these to be more static and repeatable by converting them to `var`'s instead of `val`'s thereby constraining each execution to minimize unintended leakage.